### ISSUE TITLE
Debug checks that would be nice to optimize or provide a way to turn off

### ISSUE THREAD

**Issue by jacob314** [flutter/flutter#76263]
## Use case
Flutter debug mode performance needs to be fast enough that users can understand how their app feels without frequently requiring a profile build. The difference in productivity iterating on an app in a profile build with Flutter is much lower than with a debug build due to hot reload.
Slow performance in debug builds is mainly due to useful asserts in the Flutter framework that help catch bugs but can get in the way when users are using debug builds to run to see how their app feels instead of to debug issues.

Some ideas:
* Optimize some of the slowest asserts. A few of the asserts take up a significant amount of time when profiling the Flutter DevTools app itself.
* Provide a way to run debug builds with some or all asserts disabled (the Dart VM has an --enable-assert flag to toggle asserts but it isn't exposed through Flutter). This would give users who want a fast hot-reload iteration cycle while experimenting in the UI the fastest debug build performance at the risk of missing the most commonly found Flutter errors.

Exactly which asserts contribute time depends on the app and I haven't seen as consistent results as I would like between profiling Flutter DevTools itself and profiling the Flutter Gallery. This is likely as many asserts walk up the tree of Elements or RenderObjects looking for a match. Certain apps may result in quite long tree walks having a much larger performance penalty. 

Expensive debug checks:
These debug checks mostly share the property that in the worst case they walk far up the tree looking for a match. 
- [ ] _debugCanPerformMutations  https://github.com/flutter/flutter/issues/76261
- [ ] _debugIsInScope shows up sometimes but not in every profile.
- [ ] _debugCheckForCycles https://github.com/flutter/flutter/issues/76259
- [ ]  debugCheckHasDirectionality https://github.com/flutter/flutter/issues/76262

- [ ] TickerProviderStateMixin triggers calls to an expensive toString https://github.com/flutter/flutter/issues/76257

**Comment by jacob314** [comment]
Fyi @Hixie

**Comment by jacob314** [comment]
@nshahan has been doing some experiment with Flutter Web that show just how large the impact of these asserts is on the web.

**Comment by Hixie** [comment]
I assigned this to myself optimistically, but realistically I'm not going to be able to get to this for a while, so everyone else should feel empowered to take it if they are motivated to work on it.

**Comment by jonahwilliams** [comment]
FWIW I hooked up a basic example of disabling some of these asserts with a VM Service toggle here: https://github.com/flutter/flutter/pull/79714

I did not notice a difference in either Flutter gallery

**Comment by Hixie** [comment]
We should definitely revisit each of these to see if we can optimize them, see if there's a reason we might not need them anymore (e.g. null safety making them redundant), etc, but any turning off of these asserts, even if they are measurably impacting performance, should be done on a case-by-case basis with a lot of care and thought. They are pretty important.

**Comment by jonahwilliams** [comment]
Agreed, but for the sake of the discussion I wanted to see how much of an effect just turning them off would have. Conclusion: 🤷🏻‍♂️

**Comment by Hixie** [comment]
Yeah sorry I didn't mean to imply the test was a bad idea. :-) I'm surprised it had minimal impact. Might be dependent on the kind of load we apply...

**Comment by Hixie** [comment]
I looked at three of these and submitted PRs for them. There's still two more.

**Comment by cj-radcliff** [comment]
Closing as outdated, please reopen if necessary
